scanner/python.go

```diff
@@ -1,8 +1,261 @@
 package scanner
 
-func configurePython(sourceDir string, config *ScannerConfig) (*SourceInfo, error) {
-	// using 'poetry.lock' as an indicator instead of 'pyproject.toml', as Paketo doesn't support PEP-517 implementations
-	if !checksPass(sourceDir, fileExists("requirements.txt", "environment.yml", "poetry.lock", "Pipfile", "setup.py", "setup.cfg")) {
+import (
+	"bufio"
+	"os"
+	"path/filepath"
+	"strings"
+	"unicode"
+
+	"github.com/superfly/flyctl/terminal"
+	"golang.org/x/exp/slices"
+)
+
+type PyApp string
+
+const (
+	FastAPI   PyApp = "fastapi"
+	Flask     PyApp = "flask"
+	Streamlit PyApp = "streamlit"
+)
+
+var supportedApps = []PyApp{FastAPI, Flask, Streamlit}
+
+type PyProjectCfg struct {
+	pyVersion     string
+	appName       string
+	supportedApps []PyApp
+}
+
+func findEntrypoint(dep string) *os.File {
+	var entrypoint *os.File = nil
+	filepath.Walk(".", func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if filepath.Ext(path) == ".py" && !strings.Contains(path, ".venv") {
+			file, err := os.Open(path)
+			if err != nil {
+				return err
+			}
+			defer file.Close() // skipcq: GO-S2307
+
+			scanner := bufio.NewScanner(file)
+			for scanner.Scan() {
+				line := scanner.Text()
+				if strings.Contains(line, "import") && strings.Contains(line, dep) {
+					entrypoint = file
+				}
+			}
+
+			if err := scanner.Err(); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	return entrypoint
+}
+
+func parsePyDep(dep string) string {
+	// remove all version constraints from a python dependency
+	// e.g. "fastapi>=0.1.0" -> "fastapi"
+	// e.g. "flask" -> "flask"
+	// e.g. "pytest < 5.0.0" -> "pytest"
+	// e.g. "numpy~=1.19.2" -> "numpy"
+	// e.g. "django>2.1; os_name != 'nt'" -> "django"
+	dep = strings.ToLower(dep)
+	dep = strings.Split(dep, ";")[0]
+	dep = strings.Split(dep, " ")[0]
+	dep = strings.Split(dep, "==")[0]
+	dep = strings.Split(dep, ">")[0]
+	dep = strings.Split(dep, "<")[0]
+	dep = strings.Split(dep, "~=")[0]
+	return dep
+}
+
+func fromPoetry(pyProject map[string]interface{}) (PyProjectCfg, error) {
+	// Parse pyproject.toml managed with poetry
+	deps := pyProject["tool"].(map[string]interface{})["poetry"].(map[string]interface{})["dependencies"].(map[string]interface{})
```

Comment:
I may be needlessly worrying (and feel free to ignore me if so), but this series of type assertions concern me.  If the toml file does not contain this exact series of nested maps/dictionaries, flyctl will panic and halt:

https://go.dev/ref/spec#Type_assertions

As annoying as it might be, you might want to code this as a series of assignment statements, and return an error if any of the assignment statements returns a not ok.

Reply:
I got some additional advice to rewrite this into structs and use the unmarshal feature. I'll implement that and cleanup the code a bit more - I think the result will be a bit more robust; thanks for the feedback!